c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -5451,6 +5474,437 @@ bool constructor(JSContext *cx, unsigned argc, Value *vp) {
+    JS_ReportErrorASCII(cx, "ObjectStore key can not be \"..\"");
+    return false;
+  }
+
+  HandleValue body_val = args.get(1);
+
+  // We currently support five types of body inputs:
+  // - byte sequence
+  // - buffer source
+  // - USV strings
+  // - URLSearchParams
+  // - ReadableStream
+  // After the other other options are checked explicitly, all other inputs are
+  // encoded to a UTF8 string to be treated as a USV string.
+  // TODO: Support the other possible inputs to Body.
+
+  RootedObject body_obj(cx, body_val.isObject() ? &body_val.toObject() : nullptr);
+
+  if (body_obj && JS::IsReadableStream(body_obj)) {
+    JS_ReportErrorLatin1(cx, "Streaming into an ObjectStore is not yet supported");
+    return false;
+  } else {
+    mozilla::Maybe<JS::AutoCheckCannotGC> maybeNoGC;
+    UniqueChars text;
+    char *buf;
+    size_t length;
+
+    if (body_obj && JS_IsArrayBufferViewObject(body_obj)) {
+      // Short typed arrays have inline data which can move on GC, so assert
+      // that no GC happens. (Which it doesn't, because we're not allocating
+      // before `buf` goes out of scope.)
+      maybeNoGC.emplace(cx);
+      JS::AutoCheckCannotGC &noGC = maybeNoGC.ref();
+      bool is_shared;
+      length = JS_GetArrayBufferViewByteLength(body_obj);
+      buf = (char *)JS_GetArrayBufferViewData(body_obj, &is_shared, noGC);
+    } else if (body_obj && JS::IsArrayBufferObject(body_obj)) {
+      bool is_shared;
+      JS::GetArrayBufferLengthAndData(body_obj, &length, &is_shared, (uint8_t **)&buf);
+    } else if (body_obj && URLSearchParams::is_instance(body_obj)) {
+      SpecSlice slice = URLSearchParams::serialize(cx, body_obj);
+      buf = (char *)slice.data;
+      length = slice.len;
+    } else {
+      // Convert into a String following https://tc39.es/ecma262/#sec-tostring
+      text = encode(cx, body_val, &length);
+      // 30MB in bytes is the max size allowed for ObjectStore.
+      if (length > 30 * 1024 * 1024) {
+        JS_ReportErrorLatin1(cx, "ObjectStore value can not be more than 30 Megaytes in size.");
+        return false;
+      }
+      if (!text) {
+        return false;
+      }
+      buf = text.get();
+    }
+
+    BodyHandle body_handle = {INVALID_HANDLE};
+    if (!HANDLE_RESULT(cx, xqd_body_new(&body_handle))) {
+      return false;
+    }
+
+    if (body_handle.handle == INVALID_HANDLE) {
+      return false;
+    }
+
+    int result = write_to_body_all(body_handle, buf, length);
+
+    // Ensure that the NoGC is reset, so throwing an error in HANDLE_RESULT
+    // succeeds.
+    if (maybeNoGC.isSome()) {
+      maybeNoGC.reset();
+    }
+
+    if (!HANDLE_RESULT(cx, result)) {
+      return false;
+    }
+
+    size_t inserted = 0;
+    // max-age is ignored for Object Store
+    uint32_t max_age = 0;
+
+    int status = fastly_kv_insert(object_store_handle(self), key_chars, key_len, body_handle,
+                                  max_age, &inserted);
```

Comment:
This is the now an outdated hostcall, once the new ones are released I can update this pull-request to use the new ones